[CIR] Upstream lowering of conditional operators to TernaryOp

mmha opened the pull request:
This patch adds visitors for BinLAnd, BinLOr and AbstractConditionalOperator. Note that this patch still lacks visitation of OpaqueValueExpr which is needed for the GNU ?: operator.

mmha commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -948,6 +950,165 @@ void CIRGenFunction::emitIgnoredExpr(const Expr *e) {
+  SmallVector<mlir::OpBuilder::InsertPoint, 2> insertPoints{};
+  mlir::Type yieldTy{};
+
+  auto emitBranch = [&](mlir::OpBuilder &b, mlir::Location loc, Expr *expr,
```
I changed the control flow here wrt. to the incubator. Note that I patch up the regions after the `builder.create` instead of doing this within the false expression builder. I *think* that's okay but I rather highlight this.

erichkeane reviewed (commented):
Whew... this ends up being a sizable patch.  I know you tried to break it up, but it is still pretty significant.

I have a couple of comments as drive-bys, else I'll let others do the approvals.

mmha committed:
Update clang/lib/CIR/CodeGen/CIRGenExpr.cpp

Co-authored-by: Erich Keane <ekeane@nvidia.com>

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -1658,6 +1828,170 @@ mlir::Value ScalarExprEmitter::VisitUnaryExprOrTypeTraitExpr(
+static bool isCheapEnoughToEvaluateUnconditionally(const Expr *e,
+                                                   CIRGenFunction &cgf) {
+  // Anything that is an integer or floating point constant is fine.
+  return e->IgnoreParens()->isEvaluatable(cgf.getContext());
```
We wouldn't want to speculatively evaluate constant foldable floating point expressions if they might raise an exception. I don't think this handles that correctly. We are a long way from having strict floating-point semantics in CIR, but the classic codegen gets this wrong: https://godbolt.org/z/nvfT4WxK4

mmha replied:
Good catch! I modified your example a bit because of a comment in `Expr.h` on `Expr::EvalStatus::HasUndefinedBehavior`: https://godbolt.org/z/Kvn3M9hr7

> For example, 1.0 / 0.0 can be folded to Inf, but has undefined behavior.

Looks like Info.noteUndefinedBehavior() doesn't get called for FP division by zero in `handleFloatFloatBinOp()`. Easy enough fix, I'll create a separate PR for this :) 

If we want to be strict about this wouldn't this exclude almost any speculative FP expression evaluation as the FP environment can be set at runtime to make this observable? E. g. trapping on inexact results happens pretty much all the time.

mmha replied:
FWIW even if `isCheapEnoughToEvaluateUnconditionally` returns true we still emit a `TernaryOp` instead of a `SelectOp`. This needs to be fixed here and in the incubator.

andykaylor replied:
I guess this is really a matter of isEvaluatable() giving us a bad answer. It should be taking FP exceptions into account when FP exceptions are enabled.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -875,6 +877,174 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    // If we have 0 && RHS, see if we can elide RHS, if so, just return 0.
+    // If we have 1 && X, just emit X without inserting the control flow.
+    bool lhsCondVal;
+    if (cgf.constantFoldsToSimpleInteger(e->getLHS(), lhsCondVal)) {
```
It feels like we have code the code to check for this folding and then create a ternary op in a lot of places. Is there no way to reduce the duplication?

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -948,6 +950,165 @@ void CIRGenFunction::emitIgnoredExpr(const Expr *e) {
+  return info;
+}
+
+LValue CIRGenFunction::emitConditionalOperatorLValue(
```
It doesn't look like this is called from anywhere in this patch. In the incubator, it is called from emitLValue(), and the fact that you didn't need that here makes me suspect this is entirely duplicated somewhere else (maybe in the VisitAbstractConditionalOperator implementation?).

andykaylor reviewed (commented):
Could you break this up a bit more? Maybe cut it back to just the code needed for the logical operators?

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -875,6 +877,174 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+                                                b.getInsertionBlock()};
+          cgf.curLexScope->setAsTernary();
+          mlir::Value rhsCondV = cgf.evaluateExprAsBool(e->getRHS());
+          auto res = b.create<cir::TernaryOp>(
```
I think this second ternary is entirely unnecessary. This code is effectively doing this for `a && b`:
```
t1 = evaluateExprAsBool(a);
if (t1) {
  t2 = evaluateExprAsBool(b);
  if (t2) {
    result = true;
  } else {
    result = false;
  }
} else {
  result = false;
}
```
Couldn't the same result be accomplished like this?
```
t1 = evaluateExprAsBool(a);
if (t1) {
  result = evaluateExprAsBool(b);
} else {
  result = t1;
}
```
Similar reasoning applies to the LOr handling below.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -875,6 +877,174 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+          auto res = b.create<cir::ConstantOp>(loc, builder.getFalseAttr());
+          b.create<cir::YieldOp>(loc, res.getRes());
+        });
+    return builder.createZExtOrBitCast(resOp.getLoc(), resOp.getResult(),
```
VisitUnaryLNot does this:
```
  // ZExt result to the expr type.
  mlir::Type dstTy = cgf.convertType(e->getType());
  if (mlir::isa<cir::IntType>(dstTy))
    return builder.createBoolToInt(boolVal, dstTy);
  if (mlir::isa<cir::BoolType>(dstTy))
    return boolVal;

  cgf.cgm.errorNYI("destination type for logical-not unary operator is NYI");
```
A helper function seems reasonable, but I'd prefer to see it be named something like `maybePromoteBoolResult`.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -875,6 +877,174 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    mlir::Location loc = cgf.getLoc(e->getExprLoc());
+
+    // If we have 0 && RHS, see if we can elide RHS, if so, just return 0.
+    // If we have 1 && X, just emit X without inserting the control flow.
```
This results in inconsistent behavior. We generate a pair of `cir.ternary` ops for `x && true` but we simplify `true && x`. 

https://godbolt.org/z/1rfYT8E46

It seems to me that it would be better to leave both alone and let the optimizer do the optimization.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenFunction.h:
```diff
@@ -642,12 +798,97 @@ class CIRGenFunction : public CIRGenTypeCache {
+        : insertPt(cgf.builder.saveInsertionPoint()) {}
+    ConditionalEvaluation(mlir::OpBuilder::InsertPoint ip) : insertPt(ip) {}
+
+    void begin(CIRGenFunction &cgf) {
```
If we saved a reference to `cgf` in the constructor, we wouldn't need to pass it to `begin()` and `end()`

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -948,6 +950,165 @@ void CIRGenFunction::emitIgnoredExpr(const Expr *e) {
+      if (mlir::isa<cir::VoidType>(yieldTy)) {
+        builder.create<cir::YieldOp>(loc);
+      } else { // Block returns: set null yield value.
+        mlir::Value op0 = builder.getNullValue(yieldTy, loc);
```
I don't think this is right. If a block contains a throw, the incubator generates a `cir.throw` followed by a `cir.unreachable` and then inserts a null constant and a `cir.yield`. This causes verification to fail because `cir.unreachable` isn't the last operation in the block.

mmha replied:
https://github.com/llvm/clangir/issues/1647

mmha committed:
typo fix

mmha commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -948,6 +950,165 @@ void CIRGenFunction::emitIgnoredExpr(const Expr *e) {
 
+// Handle the case where the condition is a constant evaluatable simple integer,
+// which means we don't have to separately handle the true/false blocks.
+static std::optional<LValue> handleConditionalOperatorLValueSimpleCase(
```
Do you want to move this to a CIR pass eventually? The downside of that would be a bunch of extra dead code being generated.

andykaylor replied:
I would prefer to see it happen in CIRSimplify. Especially in the case of a ternary operator, I wouldn't expect it to be common for there to be a lot of dead code.

mmha committed:
Address review feedback

mmha committed:
Account for aligned load/stores

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenFunction.cpp:
```diff
@@ -201,6 +201,20 @@ bool CIRGenFunction::constantFoldsToSimpleInteger(const Expr *cond,
+/// If the specified expression does not fold
+/// to a constant, or if it does but contains a label, return false.  If it
+/// constant folds return true and set the boolean result in `resultBool`.
+bool CIRGenFunction::constantFoldsToSimpleInteger(const Expr *cond,
```
I believe we already have this function, but it was renamed to `constantFoldsToBool`

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenFunction.h:
```diff
@@ -108,6 +108,152 @@ class CIRGenFunction : public CIRGenTypeCache {
+  // multiple VLA types can share the same size expression.
+  // FIXME: Maybe this could be a stack of maps that is pushed/popped as we
+  // enter/leave scopes.
+  llvm::DenseMap<const Expr *, mlir::Value> vlaSizeMap;
```
Is this used in this PR?

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -1799,6 +1870,162 @@ mlir::Value ScalarExprEmitter::VisitUnaryExprOrTypeTraitExpr(
+  ignoreResultAssign = false;
+
+  // Bind the common expression if necessary.
+  CIRGenFunction::OpaqueValueMapping binding(cgf, e);
```
This doesn't seem to be necessary yet. As far as I can tell, we'll need this when we add handling for OpaqueValueExprClass but not until then.

andykaylor commented on clang/test/CIR/CodeGen/binop.cpp:
```diff
@@ -540,3 +540,171 @@ void long_shift_example(long long a, short b) {
+// LLVM: [[OR_CONT]]:
+// LLVM: %[[ZEXT_OR:.*]] = zext i1 %[[OR_PHI]] to i8
+// LLVM: store i8 %[[ZEXT_OR]], ptr %[[X]]
+// LLVM: ret void
```
Can you add OGCG checks for comparison?

andykaylor reviewed (commented):
This looks good. My remaining concerns are mostly about dead code.

mmha committed:
[CIR] Upstream lowering of lvalue conditional operators to TernaryOp

This patch adds visitors for BinLAnd, BinLOr and AbstractConditionalOperator. Note that this patch still lacks visitation of OpaqueValueExpr which are needed for the GNU ?: operator.

mmha committed:
Review feedback: Remove dead code, add OGCG tests

andykaylor reviewed (approved):
lgtm

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -98,6 +98,14 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+      return builder.createBoolToInt(value, dstTy);
+    if (mlir::isa<cir::BoolType>(dstTy))
+      return value;
+  }
```
Note the 'else' case here is problematic.  We need a return here.
